### DISCUSSION TITLE
2023 roadmap and issues cleanup

### DISCUSSION THREAD

**Discussion by katspaugh** [katspaugh/wavesurfer.js#2684] [Announcements]
Dear wavesufer.js community 👋 ,

I've decided to do some spring cleaning and closed most of the issued we had open. 🧹
90% of these issues were from more than a year ago, and most of them were questions/dev support requests that were either answered, or never followed up on by their OP.

Instead of creating issues with questions, we encourage the community to use the [Q&A section](https://github.com/wavesurfer-js/wavesurfer.js/discussions/categories/q-a) in this forum. I've migrated some recent questions there.

There was also a sizable number of feature ideas which should now be created in the [Ideas forum section](https://github.com/wavesurfer-js/wavesurfer.js/discussions/categories/ideas).

## Better documentation
The amount of dev support requests is a sign that we need to make our documentation more accessible.

* Provide easy-to-follow tutorials for common use cases
* Write interactive docs with live code snippets ([storybook](https://storybook.js.org) maybe?)
* Reorganize the automatically generated API docs
* Perhaps create video tutorials (or collect existing ones if any)

## TypeScript
Another thing I propose to tackle in 2023 is migrating the project to TypeScript. 
This will help app developers discover methods/events and their signatures w/o going to the documentation. And it will reduce the possibility of trivial bugs both in wavesurfer.js's code itself and the code that uses it.

## Feedback
Your feedback and ideas are very welcome! 🏄

**Comment by thijstriemstra** [comment]
it's quite shocking to wake up to 600 emails from wavesurfer project, having no idea what's going on.

> I've decided to do some spring cleaning and closed most of the issued we had open.

isn't this something we should've discussed among everyone first? i'm a bit speechless atm.

**Reply by katspaugh** [reply]
True, sorry, I should've put it up for discussion first.

**Reply by katspaugh** [reply]
@thijstriemstra btw one of those emails was an invitation to the new org, I hope it wasn’t lost in the crowd. 😅

**Reply by entonbiba** [reply]
@thijstriemstra lol you don't organize your emails in folders 

@katspaugh it's good you did that clean up most of the issues are very outdated. I like the whole discussions section setup.

**Reply by thijstriemstra** [reply]
> lol you don't organize your emails in folders

I did?

**Reply by katspaugh** [reply]
I guess it's less about the emails themselves and more about someone (me) suddenly doing such a drastic change w/o consulting with the main maintainer and the community. I'll do my best to first propose and discuss things in the future.

**Comment by katspaugh** [comment]
@thijstriemstra @entonbiba would you guys be up for a call sometime this week to align on a few things?

Namely
* the roadmap
* dev support strategy
* funding (organizational questions)

**Reply by entonbiba** [reply]
@katspaugh for sure, I created a private project for contact info let me know if you got the invite there?

**Reply by thijstriemstra** [reply]
In order to document it I'd prefer to do a text chat somewhere, let me know what works for you (discord?).

**Reply by katspaugh** [reply]
I will take notes in a notion doc, that's not a problem. But I understand if you're not comfortable with a call.

**Reply by entonbiba** [reply]
yea notes wise is not a problem, but we should do a call or video

**Reply by prakharpbuf** [reply]
wayyy out of my line but if all of you are okay with it ...can I join?
I might not have much (or any) useful input but I'll listen and learn.

**Reply by entonbiba** [reply]
@prakharpbuf this is primarily for contributors and to see how best we can support it going forward by streamlining the strategy, for sure we will update the roadmap after with our findings.

If you have any pull requests you can contribute to the project that would be great and helpful as well!

**Reply by sundayz** [reply]
@thijstriemstra since you suggested Discord, maybe starting a Discord community server would make it easier to organize information and keep up with everything?

**Reply by corysimmons** [reply]
I'd love to nerd out about Wavesurfer and Tone.js (and just WebAudio in general) if anyone wants to start one up.

**Comment by sundayz** [comment]
If you're going to be rewriting this to typescript, I suggest you also clean up the audio backend class hierarchy. Right now it is `Observer > WebAudio > MediaElement > MediaElementWebAudio` which doesn't make sense, because MediaElement is not related to WebAudio. Instead, there should be an interface that all these classes fulfill in their own way, e.g.:

```ts
interface AudioBackend {
  play(): Promise<void>;
  pause(): Promise<void>;
  getVolume(): number;
 // ...
}

class WebAudio implements AudioBackend {
  // ...
}

class MediaElement implements AudioBackend {
  // ..
}
```

it's a big task, however :^)

**Reply by katspaugh** [reply]
Good point!

**Reply by katspaugh** [reply]
I've made a little [diagram](https://kinopio.club/wavesurfer-js-architecture-UlM_1jklJtbzkyF8T-yG5) of how I think it should be:

<img width="832" alt="Screenshot 2023-03-02 at 07 33 13" src="https://user-images.githubusercontent.com/381895/222349436-38b550e5-24dc-4143-9cdb-efbe00540213.png">

Apart from making MediaElement and WebAudio equal citizens, I think it would be nice to extract the Decoding and Fetching logic in their own respective modules.

**Reply by katspaugh** [reply]
I've started an experimental rewrite here: https://github.com/katspaugh/wavesurfer.ts
Not sure rewriting from scratch is the best way to go, but it puts things in a refreshing perspective.
We should discuss possible approaches.

**Comment by kai-shimada** [comment]
Hello, I think it is a great idea to rewrite wavesurfer.js and at the same time, maybe it's a great opportunity for me to give you just a source to improve rendering and zooming performance for the new wavesufer!

Maybe you are already aware of this repo, but if not, they've made a great example of how smooth the waveform rendering can be, I haven't seen it anywhere else (sadly its unmaintained for about 8 years but it still performs). 

Here is the repo: https://github.com/wavesjs/waves-ui
Here are the examples they've made: http://wavesjs.github.io/waves-ui/manual/example.html#examples

Just look at the `Zoom States` example, it's very smooth. If you make like one overview peaks with lower resolution, and make a switch to full resolution when zooming in at some point, it will perform very well even on long audio files. Since the wavesurfer has great support for other features, it might be a great way to really solve the wavesurfer.js rendering/zooming performance difficulties and make wavesurfer.js a great overall solution.

**Reply by katspaugh** [reply]
Thank you! I haven't seen this lib, very smooth zooming indeed!
They're only rendering the part visible in the viewport, which makes it much faster than what we do. When moving along the waveform, or zooming, it re-renders within the same visible "window". The approach wavesurfer.js took was to have an actual scrollbar.
I've [optimized](https://github.com/katspaugh/wavesurfer.ts/blob/main/src/renderer.ts#L200-L220) this in the new version by delaying the rendering of invisible parts, so it should be already pretty fast. Try high levels of zoom here: https://wavesurfer-ts.pages.dev/tutorial/#/examples/zoom.js
I think an actual scroll has better accessibility and possibly UX, and definitely simpler in terms of code. But the windowing/panning approach is still on the table, the performance is great.

**Reply by kai-shimada** [reply]
Ok I see. I just tried your example and it is indeed performing much better!
I know it might sound nitpicky, but the performance degrades with increasing zoom value, It works better in chrome but worse in safari especially if you jump from zoomed-in to zoomed-out view.

Chrome:

https://user-images.githubusercontent.com/79719366/226917759-39d1f75c-33c0-4e04-805a-72e87fcb205b.mov


Jump back and forth in Safari:

https://user-images.githubusercontent.com/79719366/226917844-6e629336-19ad-44d1-8336-c57e0d040533.mov

**Reply by katspaugh** [reply]
Thank for taking the time to test and record it!
Safari is pretty slow indeed. It can be definitely optimized further.
Update: performance on Safari is now fixed.

**Reply by kai-shimada** [reply]
I tested it again now, and not only it is working better on safari (safari still seems to underperform compared to chrome), the performance on chrome got even better and the micro hiccups are all gone - impressive work and I really appreciate how fast it develops - is there a approx. timeline on when there will be a "stable" release?

**Reply by katspaugh** [reply]
I’m still getting occasional bug reports for the alpha, but am planning to release a beta likely this weekend. I’ll keep you all posted.

**Reply by entonbiba** [reply]
@katspaugh awesome! :)
